apis/stargate/v1alpha1/stargate_types.go

```diff
@@ -158,16 +162,19 @@ type StargateRackTemplate struct {
 	Name string `json:"name"`
 }
 
-// Coalesce compares this StargateRackTemplate with the given StargateDatacenterTemplate and returns the first non-nil
-// StargateTemplate it finds.
-// TODO revisit the merging strategy and/or find a better name for this method
-func (in *StargateRackTemplate) Coalesce(dcTemplate *StargateDatacenterTemplate) *StargateTemplate {
+// Merge merges the given StargateDatacenterTemplate with this StargateRackTemplate and returns the
+// result.
+func (in *StargateRackTemplate) Merge(dcTemplate *StargateDatacenterTemplate) (*StargateTemplate, error) {
```

Comment:
> So you might have a method signature: func (in *StargateDatacenterTemplate) Merge(rackTemplate *StargateRackTemplate) (*StargateTemplate, error) instead of this one to make that structure clearer?

Currently we have two merge methods where the semantics are: the receiver receives a template and apply non-zero values from that template to zero values in itself, and returns the result.

What you suggest would be: the receiver receives a template and applies its non-zero values to all of the template's zero values, and returns the result.

I don't have a strong opinion here, but what we have seems to me slightly more straightforward that the other way around.

> do we want to define Merge methods on these types? It would be nice if we could avoid defining any methods at all

Yes, I am wondering myself. For Cassandra types, we don't have methods currently, but a simple utility function. I can do the change if we agree that a function is better than methods on API types. (For the nilness, we'd still need to check that before passing the objects to Mergo).

> On a related note, IMO a merge function should always have function signature func (a Type1) Merge (b Type1) (Type1, error) whereas here we have func (a Type1) Merge (b Type2) (Type3, error), which isn't great for consistency with the existing codebase and might get us confused in the long term.

I agree on the principle but Stargate types have a complex hierarchy in fact. I will see if the hierarchy itself could be simplified somehow, but that might lead us out of the scope of this PR.

Also, tbh, I think the most idiomatic behavior in Go for a Merge method would be to mutate the receiver in-place, not return the merge result. That also advocates for a utility function rather than a method, in fact.

Reply:
> Also, tbh, I think the most idiomatic behavior in Go for a Merge method would be to mutate the receiver in-place, not return the merge result. That also advocates for a utility function rather than a method, in fact.

Yes, we should follow the structure that Kubernetes uses with the `DeepCopyInto` instead of creating a merged copy.

Reply:
> I can do the change if we agree that a function is better than methods on API types.

Done in d56b14a, let me know what you think @Miles-Garnsey.

Reply:
A few things here:

### Most recent commit

The most recent commit is looking a fair bit better to me and is definitely a clearer read. I do have some comments that I've put inline.

### Can we override all (nil + non-nil) values?

The merge [methods](https://github.com/Miles-Garnsey/k8ssandra-operator/blob/feature/create-medusa-mapping/apis/telemetry/v1alpha1/telemetry_methods.go#L4) I've been working on are of the form `func (a T) Merge (b T) T` where all non-nil values defined in `b` overwrite all (nil **and non-nil**) values in `a`. I've just remembered that Mergo doesn't follow this overwriting behaviour unless you add `WithOverride` as an option. Is that something you'd be open to doing?

IMO, we want to treat every `a` value as a base and apply successive refinements from `b`/right hand objects (e.g. cluster -> DC -> rack) which overwrite any default values inherited from the more general object to the left.

The way I see this being used, users will define a set of cluster-wide defaults, but may want to override those defaults for a single DC/rack (e.g. they spin up a new DC which is on-prem, or on a different cloud). In the current logic, the presence of a cluster-level default appears to block the user from being able to specify a special option for that unique DC. (Am I reading this right?)

(And I apologise for missing this before, I realise it is a big change. I'm still feeling my way around this problem.)

### Zero vs nil values

> apply non-zero values from that template to zero values in itself

Nit: you mean non-nil and nil here I think. Zero values are `""`, `0`, etc. and would get applied. Just making sure we're on the same page.

### Pointers, purity etc.

I think the function should remain pure. There's no reason to mutate inputs here and it makes the code harder to reason about.

As it stands, the function is pure, but the signature is a little misleading. When you take pointer inputs, it leads the reader to the assumption that you'll mutate the objects in-place. It is a bit odd to then see a pointer as an output. It might be good for clarity to just avoid the use of pointers here.

(I also acknowledge that I was passing pointers around in my own `Merge` function, so I'm guilty  of setting a bad example here.)

### Methods vs functions

When I asked whether we had to define these methods, I was actually wondering if we couldn't call `mergo.Merge()` directly. As you say, this doesn't work if either input is nil, so we do need to wrap it.

I have no strong preference between methods/functions, what you have looks good to me. I also very much appreciate the way that Merge inputs/outputs a single type now. Much clearer!

I did idly wonder if there is a way that we can make this function take an `interface()` so we can avoid defining it for every type. I don't mind either way since once generics are available we can probably safely use them for this task. No need to make any further changes for now though.

Reply:
On reflecting more on this, I've realised my commentary under `Can we override all (nil + non-nil) values?` may be a subjective preference, and not an essential change. I was assuming a right biased operation (i.e. values on the right override all values on the left) but because Mergo's `WithOverride` is not set, it actually means that the operation is left biased (values are only merged from the right if they are unset on the left, so the left takes precedence). 

I've left the commentary there because (a) it explains why we might have had some disconnects in our communication and (b) it is an interesting reflection of the psychological complexities that arise when your colleagues are European/American and drive on the wrong side of the road.

Reply:
@adutra I've just edited the above and added some discussion on where the overriding of `false` values with `true` might become a problem. Not sure how we'll get around that one.

Reply:
I'd also like to point out that, for the sake of merging cluster-level values into dc-level ones, we aren't going to need `WithOverride`.

Keeping in mind that a cluster-level value should apply to each DC, **_unless the DC re-defines that value_**, `WithOverride` is NOT what we want and would make cluster-level values "win" over DC-level ones.

Reply:
One last thing @Miles-Garnsey and sorry for the noise. In a comment that I can't find anymore, you referred to [these methods](https://github.com/k8ssandra/k8ssandra-operator/blob/d728668b76cf1e9da2eeb78842ecfb2c7cade49f/apis/telemetry/v1alpha1/telemetry_methods.go#L4).

I wasn't aware of those, my apologies. Indeed they are good candidates for Mergo. Do you want me to change them as well? Judging from the implementation though, it seems you designed them so that the passed argument overwrites the receiver. I think the same could be achieved with Mergo using `WithOverwriteWithEmptyValue`.

Reply:
> If the default behavior (true replaces false) is not desired for a given field, then it is a sign that that field should rather be a *bool.

I agree with this, but it doesn't solve our problem where a user defines things at a cluster level but has one 'odd' DC or rack that must be treated differently and which they want to explicitly set false.

Even if the value is a `*bool`, a a `*true` defined at the cluster level will still override a `*false` at the rack/DC level.

The problem with this logic is that if a value is set true anywhere in the cluster or DC levels, the lower DC/rack levels can't ever override it. I do not think that is the behaviour we want, is it?

> But then this wouldn't be a merge anymore, would it? You would be effectively skipping the field from merge.

If you look at my Merge logic from the original methods I wrote [here](https://github.com/k8ssandra/k8ssandra-operator/blob/d728668b76cf1e9da2eeb78842ecfb2c7cade49f/apis/telemetry/v1alpha1/telemetry_methods.go#L4) you'll see that the behaviour differs from what we have here in two ways:

1. Fields from b are always merged into a if they are non-nil in b. This will happen even if they are non-nil (AND non-zero) in a. All values defined in b override those defined in a. 
2. The logic doesn't hinge on zero values, whether a value is zero or not, if a value in b is non-nil, it will override the value in a.

For a concrete example of how this might be a problem, my Stargate telemetry data structures and the current Mergo behaviour is going have the following behaviour:
1. A user has set up 2 DCs and both have Prometheus enabled for Stargate at cluster level. 
4. They go to set up another DC where Prometheus isn't available (I don't know why, maybe they're using a different solution in that DC).
5. They will set Prometheus = `false` in that DC. But that setting will be overwritten with `true` 🙀

Reply:
Also, I've done some digging and I'm uncertain whether `WithOverwriteWithEmptyValue` will work to avoid the issues I'm raising. When I use it [here](https://go.dev/play/p/CvYUjVMIULE) I'm seeing behaviour where the lack of a value in s2 overwrites s1's Field2 with nil.

This isn't what we want, we're trying to implement proper 3 value logic here for bools where a nil is always overwritten, but which value wins is decided by whether it is on the left or the right, not by whether it is true or false.

Reply:
I've replied offline already but I'm repeating my thoughts here for visibility.

The default merge behavior is IMO the right one.

For boolean fields, the default merge behavior results in the following truth table:

| Cluster-level value  | DC-level value | Merge Result | Comments |
| ------------- | ------------- | ------------- | ------------- |
| false  | false | false | |
| false  | true | true | DC-level explicit opt-in |
| true  | false | true | Cluster-level opt-in (imposed on all DCs) |
| true  | true | true | Cluster- and DC-level explicit opt-in |

You'll notice that this is a standard boolean OR truth table. We disagree on row 3: you would like that to result in false (=explicit DC-level opt-out), I would like that to result in true (forced opt-in). 

As stated already, my main reason for advocating for true here is that there is no way in Go to distinguish between:

* a `false` that was set explicitly by the user to clearly express their intent to opt-out, as in `Foo{Enabled: false}`;
* and a `false` that was simply the result of the user not setting any explicit value, which could be interpreted as their intent to implicitly inherit whatever value was defined at the cluster level, as in `Foo{}`.

Due to this ambiguity, I'm in favor of letting the cluster level true value win.

You'd rightfully argue: but how can we emulate an explicit DC-level opt-out (that is, set DC level to false) if the cluster-level is set to true (opt-in)? We'd simply switch to `*bool` and a ternary logic as below:

| Cluster-level value  | DC-level value | Merge Result | Comments |
| ------------- | ------------- | ------------- | ------------- |
| nil  | nil | nil | Apply convention: true or false |
| nil  | false | false | DC-level explicit opt-out |
| nil  | true | true | DC-level explicit opt-in |
| false | nil | false | Explicit cluster-level opt-out |
| false | false | false | Cluster- and DC-level explicit opt-out |
| false | true | true | DC-level explicit opt-in overrides explicit cluster-level opt-out |
| true | nil | true | Cluster-level opt-in (imposed on all DCs) |
| true | false | true | Cluster-level opt-in (imposed on all DCs)  |
| true | true | true | Cluster- and DC-level explicit opt-in |

Note that with the above table, we have the possibility of an explicit DC-level opt-out (row 2). 

Again I suppose you would like to change row 8 to yield false. In this case indeed, the false value is not a zero value, so we know that the value was explicitly set, and that the user intent is clearly to explicitly opt-out the DC, cancelling the cluster-level opt-in. It does make some sense to me, but why bother? With row 2, we already have another way to express the situation where something is true by default (and by convention), but can be overwritten to false at DC-level.

At this point, I'm inclined to say that we need to disagree and commit, wdyt?

Reply:
Your commentary is fair. I do not believe that a `Merge()` is a boolean OR operation, because OR commutes and the order of the arguments matters for my conception of `Merge()`.

I also believe that explicit opt out should be supported. It currently is, and I'm hesitant to approve here because this is a breaking change that will modify the behaviour of existing clusters (not that I think that is a hard blocker in this case, but something to be aware of).

More importantly, because this functionality is key to how we work with hierarchies in the CR, I think it should probably be discussed amongst the broader team. I do apologise that I'm not able to provide an answer to move us forward here, please take it as a sign of how important I think this single function really is.